docs/docs/how-to/adding-common-features/adding-forms.md

```diff
@@ -5,179 +5,95 @@ examples:
 
-```jsx:title=src/pages/index.js
-import React from "react"
+Website forms have several "concerns" like data entry storage, validating fields, form submission, form storage, and other event triggers. This documentation will walk through each concern.
 
-export default function Home() {
-  return <div>Hello world!</div>
-}
-```
+Forms in Gatsby are built on top of React, so in a couple places we'll be linking to the React docs for code examples. In addition, the React ecosystem also offers some popular, well-maintained libraries like [Formik](https://github.com/jaredpalmer/formik) or [Final Form](https://github.com/final-form/react-final-form) to speed up your development process.
 
-This Gatsby page is a React component. When you want to create a form, you need to store the state of the form - what the user has entered. Convert your function (stateless) component to a class (stateful) component.
+## Form concerns
 
-```jsx:title=src/pages/index.js
-import React from "react"
+### Client-side data entry
 
-export default class IndexPage extends React.Component {
-  render() {
-    return <div>Hello world!</div>
+When users type into the input, you want to persist their input, as well as store it in a JavaScript field that will later be available to your form validation and submission logic.
+
+The "Controlled Components" section of the React documentation [provides a good code example here](https://reactjs.org/docs/forms.html#controlled-components).
+
+```
+  handleChange(event) {
+    this.setState({value: event.target.value});
   }
-}
+  ...
+   render() {
+    return (
+      ...
+      <input type="text" value={this.state.value} onChange={this.handleChange} />
+      ...
+    )
+   }
 ```
 
-Now that you have created a class component, you can add `state` to the component.
+The `handleChange` function takes the value being entered by the user and stores it internally in the component state.
 
-```jsx:title=src/pages/index.js
-import React from "react"
+### Field validation
 
-export default class IndexPage extends React.Component {
-  state = {
-    firstName: "",
-    lastName: "",
-  }
+Often you'll want to, eg, ensure that the entry in an "Email" field is a valid email address.
 
-  render() {
-    return <div>Hello world!</div>
-  }
-}
-```
+If you need to do validation of one or multiple fields, you may want to use a library like Formik, since adding logic and styling around error messages tends to be fairly boilerplate and repetitive.
 
-And now you can add a few input fields:
+### Form submission
 
-```jsx:title=src/pages/index.js
-import React from "react"
+The default behavior of the HTML form element is to send a POST request to the URL and refresh the page. You'll want to prevent the default behavior and send the information to whatever backend you're using.
```

Comment:
Probably should directly mention Gatsby Functions here (and throughout this doc) e.g. link to https://www.gatsbyjs.com/docs/reference/functions/getting-started/#forms

Reply:
Good point! Included it a couple sentences lower in the section.